torch/distributed/_composable/fsdp/_fsdp_state.py

```diff
@@ -253,3 +286,52 @@ def _get_module_fsdp_state(module: nn.Module) -> Optional[FSDPState]:
     if isinstance(state, FSDPState):
         return state
     return None
+
+
+def _register_group_forward_hooks(
+    modules: Sequence[nn.Module],
+    pre_hook: Callable,
+    post_hook: Callable,
+):
+    """
+    Registers group forward pre and post-hooks. The pre-hook runs upon the
+    first module pre-forward, and the post-hook runs upon the last. If at least
+    one module does not run forward, then the post-hook does not run.
+    """
+    modules_set = set(modules)
+    modules_to_run: Set[nn.Module] = set()
+
+    @functools.wraps(pre_hook)
+    def wrapped_pre_hook(*args: Any, **kwargs: Any):
+        in_backward = torch._C._current_graph_task_id() != -1
```

Comment:
To-do:
This is for composition with activation checkpointing. Today, (post-)forward hooks do not run in backward for activation checkpointed modules, only their forward pre-hooks. This breaks our scoreboarding and would cause forward pre-hooks to not run in the next forward.

However, `torch._C._current_graph_task_id()` does not work with `torch.compile`.

One workaround is that we know that we do not need to run the pre-forward method in backward when composing with activation checkpointing, so we could handle the logic to avoid running this in backward inside FSDP (guarded by our `TrainingState` logic).

Reply:
@awgu Can we use, the ``always_call=True`` option for the post-forward hooks? That way it will be called even with AC. I tested this in the memory tracker, when I wanted to measure post-AC memory.

Reply:
Amazing! This works :)